xonsh/foreign_shells.py

```diff
@@ -139,6 +139,8 @@ def foreign_shell_data(shell, interactive=True, login=False, envcmd='env',
         currenv = dict(currenv)
     try:
         s = subprocess.check_output(cmd, stderr=subprocess.PIPE, env=currenv,
+                                    # setsid where available to avoid hangs
+                                    preexec_fn=getattr(os, 'setsid', None),
```

Comment:
Done and much better. I was flailing about a bit on this one, but it's nice that you require Python 3 so we can use this.